### PR TITLE
@apollo/gateway: Make ApolloGateway.stop() reliable and required

### PR DESCRIPTION
**Pull request by glasser**
The (not particularly documented) ApolloGateway.stop() method didn't reliably
stop the gateway from polling. All it did was cancel a timeout. But if it is
called while pollServices is in the middle of running, it would do nothing, and
then pollServices would carry on and set another timeout.

Better semantics would be for stop() to reliably stop polling: allow the current
poll to complete, ensure that there will be no more polls, and then (async)
return. This PR implements those semantics, by implementing an explicit state
machine for ApolloGateway's polling.

One reason that these bugs were able to survive is that calling stop() is often
unnecessary. In https://github.com/apollographql/apollo-server/pull/3105 we
chose to `unref` the polling timeout to allow the Node process to exit if it's
the only thing left on the event loop, instead of encouraging users of
`ApolloGateway` to be responsible and call `stop()` themselves. While this may
be reasonable when the gateway's lifecycle is merely a schema polling timer, we
may have future changes to the gateway where proper lifecycle handling is more
important. So this PR also moves away from the world where it's fine to not
bother to explicitly stop the gateway.

That said, in the common case, we don't want users to have to write gateway
stopping code. It makes more sense for stopping the `ApolloGateway` to be the
responsibility of `ApolloServer.stop()`, just as `ApolloServer.stop()` can
trigger events in plugins. So in the recently-released Apollo Server v2.20,
`ApolloServer.stop()` calls `ApolloGateway.stop()`. This should mean that in
most cases, the missing `unref` shouldn't keep the process running, as long as
you've run `ApolloServer.stop()` (and if you don't, it's likely that other parts
of the server are keeping your process running).

What if you're still running an old version of Apollo Server? For a bit of
backwards compatibility, `ApolloGateway` detects if it appears to be connected
to Apollo Server older than v2.18. If so, it continues to do the old
unref().  If you're using v2.18 or v2.19, the upgrade to v2.20 should be pretty
painless (it's mostly just minor bugfixes). If you're using ApolloGateway on its
own for some reason, and this change causes your processes to hang on shutdown,
adding a `stop()` call should be pretty straightforward. (If we learn that this
change really is devastating, we can always go back to an unconditional
timer.unref() later.)

Fixes apollographql/apollo-server#4428.

### CODE REVIEW

**Comment by glasser** [comment]
Before merging this I should implement the Apollo Server change described in it and release v2.20. It's ready for review but I'll mark as draft for this reason.

File: gateway-js/src/index.ts [hunk 1]
```diff
@@ -243,6 +262,11 @@ export class ApolloGateway implements GraphQLService {
     apollo?: ApolloConfig;
     engine?: GraphQLServiceEngineConfig;
   }) {
+    if (this.state.phase !== 'initialized') {
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/index.ts:265]: Probably disregard, I can tell below it's explicitly written with the intention of not allowing that.

File: gateway-js/src/index.ts [hunk 2]
```diff
@@ -545,28 +596,81 @@ export class ApolloGateway implements GraphQLService {
...
     try {
       await this.updateComposition();
     } catch (err) {
       this.logger.error((err && err.message) || err);
     }
 
-    this.pollServices();
+    if (this.state.phase === 'polling') {
+      // If we weren't stopped, we should transition back to the initial 'loading' state and trigger
+      // another call to itself. (Do that in a setImmediate to avoid unbounded stack sizes.)
```
Comment (trevor-scheer) [inline gateway-js/src/index.ts:667]: Cool! TIL

**Review by trevor-scheer — APPROVED** [review]

File: gateway-js/src/index.ts (hunk 2 above)
Comment (glasser) [reply gateway-js/src/index.ts:667]: FWIW I actually don't know offhand if you get unbounded stack sizes from recursion in async functions.

Comment (glasser) [reply gateway-js/src/index.ts:667]: (I checked and you do.)

**Pull request merged**